File: theme-and-theme-variant-example/src/main/java/com/example/demo/ui/component/ImageEditor.java
package com.example.demo.ui.component;

import com.vaadin.flow.component.ClientCallable;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.button.Button;

import com.vaadin.flow.component.dependency.JsModule;

import com.vaadin.flow.component.littemplate.LitTemplate;
import com.vaadin.flow.component.notification.Notification;
import com.vaadin.flow.component.template.Id;
import com.vaadin.flow.dom.Element;


@Tag("image-editor")
@JsModule("./src/image-editor.ts")
public class ImageEditor extends LitTemplate {
	
	@Id("btnZoomIn")
	private Button zoomIn;
	
	@Id("btnZoomOut")
	private Button zoomOut;
	
	@Id("btnRotate")
	private Button rotate;
	
	@Id("btnReset")
	private Button reset;
	
	public ImageEditor() {
		
		zoomIn.addClickListener(event -> {
			getElement().executeJs("this.zoomIn()");
		});
		zoomOut.addClickListener(event -> {
			getElement().executeJs("this.zoomOut()");
		});
		rotate.addClickListener(event -> {
			getElement().executeJs("this.rotate()");
		});
		reset.addClickListener(event -> {
			getElement().executeJs("this.reset()");
		});
		
    }
	
	@ClientCallable
	public void showResetInfo() {
		System.out.println("ssss");
		Notification.show("Editor Resetado");
	}
	
}

File: custom-element-advance-example/src/main/frontend/src/image-editor.ts
import { LitElement, html, css } from 'lit';
import { customElement, property, state } from 'lit/decorators.js';

import '@vaadin/button';

@customElement('image-editor')
export class ImageEditor extends LitElement {
  @state()
  private zoom: number = 1;

  @state()
  private rotation: number = 0;

  private canvas!: HTMLCanvasElement;
  private ctx!: CanvasRenderingContext2D;
  private image!: HTMLImageElement;

  static get styles() {
    return css`
    	@import './image-editor.css';
    `;
  }
  render() {
    return html`
      <div class="editor-container">
        <canvas class="canvasEditor" width="400" height="300"></canvas>
        <div class="controls">
			<vaadin-button id="btnZoomIn">Zoom +</vaadin-button>
	        <vaadin-button id="btnZoomOut">Zoom -</vaadin-button>
	        <vaadin-button id="btnRotate">Rotate</vaadin-button>
	        <vaadin-button id="btnReset">Reset</vaadin-button>
        </div>
      </div>
    `;
  }

  firstUpdated() {
	
    this.canvas = this.renderRoot.querySelector('canvas.canvasEditor') as HTMLCanvasElement;
    this.ctx = this.canvas.getContext('2d')!;
	
    this.image = new Image();
    this.image.onload = () => this.draw();
	
    this.image.src = '/img/car.png'
	
  }

  private draw() {


	const { ctx, canvas, image, zoom, rotation } = this;

 	if (!ctx || !canvas || !image || !image.complete) {
    	console.warn("Canvas or image not ready");
    	return;
  	}

    ctx.save();
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    ctx.translate(canvas.width / 2, canvas.height / 2);
    ctx.rotate((rotation * Math.PI) / 180);
    ctx.scale(zoom, zoom);
    ctx.drawImage(image, -image.width / 2, -image.height / 2);
    ctx.restore();
  }

  private zoomIn() {
    this.zoom += 0.1;
    this.draw();
  }

  private zoomOut() {
    this.zoom = Math.max(0.1, this.zoom - 0.1);
    this.draw();
  }

  private rotate() {
    this.rotation = (this.rotation + 90) % 360;
    this.draw();
  }

  private reset() {
    this.zoom = 1;
    this.rotation = 0;
    this.draw();
	console.log("TTTTTTTTTTTT");
	(this as any).$server.showResetInfo();
  }
}
